File: ios/ComergeRuntimeBridgeValidation.swift
import Foundation

internal enum ComergeRuntimeBridgeValidation {
  static let allowedSystemTypes: Set<String> = [
    "system.auth_get_token",
    "system.auth_step_up_google",
    "system.open_app_requested",
    "system.open_shell_settings_requested",
    "system.sign_out_requested",
    "system.account_deleted",
    "system.session_invalid",
  ]

  static func isValidEnvelope(_ envelope: [String: Any], requireMicroSource: Bool) -> Bool {
    guard let v = envelope["v"] as? NSNumber, v.intValue == 1 else { return false }
    guard let type = envelope["type"] as? String, !type.isEmpty else { return false }
    guard let requestId = envelope["requestId"] as? String, !requestId.isEmpty else { return false }
    guard envelope["ts"] is NSNumber else { return false }
    guard let source = envelope["source"] as? String, (source == "host" || source == "micro") else { return false }
    if requireMicroSource && source != "micro" { return false }
    if requireMicroSource && !allowedSystemTypes.contains(type) { return false }

    let approxSize = String(describing: envelope).count
    if approxSize > 64 * 1024 { return false }
    return true
  }
}

File: android/src/main/java/expo/modules/comergeruntime/ComergeRuntimeBridgeValidation.kt
package expo.modules.comergeruntime

internal object ComergeRuntimeBridgeValidation {
  fun isValidEnvelope(envelope: Map<String, Any?>, requireMicroSource: Boolean): Boolean {
    val v = envelope["v"]
    val version = (v as? Number)?.toInt()
    val type = envelope["type"] as? String
    val requestId = envelope["requestId"] as? String
    val ts = envelope["ts"]
    val source = envelope["source"] as? String
    if (version != 1) return false
    if (type.isNullOrBlank()) return false
    if (requestId.isNullOrBlank()) return false
    if (ts !is Number) return false
    if (source != "host" && source != "micro") return false
    if (requireMicroSource && source != "micro") return false
    if (requireMicroSource && type !in ComergeRuntimeContracts.ALLOWED_SYSTEM_EVENT_TYPES) return false
    val approxSize = envelope.toString().length
    if (approxSize > ComergeRuntimeContracts.MAX_BRIDGE_PAYLOAD_CHARS) return false
    return true
  }
}
